source/translated_table_model.ts

```diff
@@ -0,0 +1,194 @@
+    if(this.transactionDepth === 0) {
+      this.dispatcher.dispatch(new Transaction(this.transactionArray));
+      this.transactionArray = null;
+    }
+    if(this.transactionDepth > -1) {
+      this.transactionDepth -= 1;
+    }
+  }
+
+  /**
+   * Moves a row.
+   * @param source - The index of the row to move.
+   * @param destination - The new index of the row.
+   * @throws RangeError - The index specified is not within range.
+   */
+  public moveRow(source: number, destination: number): void {
+    if(source >= this.rowCount || source < 0 || destination >= this.rowCount ||
+        destination < 0) {
+      throw new RangeError('The index specified is not within range.');
+    }
+    const sourceValue = this.translatedToSourceIndices[source];
+    const multiplier = source < destination ? 1 : -1;
+    for(let index = source; index != destination; index += multiplier) {
+      this.translatedToSourceIndices[index] =
+        this.translatedToSourceIndices[index + multiplier];
+      this.sourceToTranslatedIndices[
+        this.translatedToSourceIndices[index]] = index;
+    }
+    this.translatedToSourceIndices[destination] = sourceValue;
+    this.sourceToTranslatedIndices[sourceValue] = destination;
+    this.processOperation(new MoveRowOperation(source, destination));
+  }
+
+  public get rowCount(): number {
+    return this.sourceTable.rowCount;
+  }
+
+  public get columnCount(): number {
+    return this.sourceTable.columnCount;
+  }
+
+  public get(row: number, column: number): any {
+    if(row >= this.rowCount || row < 0) {
+      throw new RangeError(
+        'Row or column are outside of the table\'s bounds.');
+    } 
+    return this.sourceTable.get(this.translatedToSourceIndices[row], column);
+  }
+
+  public connect(
+      slot: (operations: Operation) => void): Kola.Listener<Operation> {
+    return this.dispatcher.listen(slot);
+  }
+
+  private processOperation(operation: Operation) {
+    if(this.transactionArray === null) {
+      this.dispatcher.dispatch(operation);
+    } else {
+      this.transactionArray.push(operation);
+    }
+  }
+
+  private handleSourceOperation = (operation: Operation) => {
+    if(operation instanceof AddRowOperation) {
+      this.sourceAdd(operation);
+    } else if(operation instanceof MoveRowOperation) {
+      this.sourceMove(operation);
+    } else if(operation instanceof RemoveRowOperation) {
+      this.sourceRemove(operation);
+    } else if(operation instanceof UpdateOperation) {
+      this.sourceUpdate(operation);
+    } else if(operation instanceof Transaction) {
+      this.beginTransaction();
+      operation.operations.forEach(this.handleSourceOperation);
+      this.endTransaction();
+    }
+  }
+
+  private sourceAdd(operation: AddRowOperation) {
+    const rowCount = this.translatedToSourceIndices.length;
+    const originalIndices = {} as {[key: number]: number};
+    for(let index = operation.index; index < rowCount; ++index) {
```

Comment:
Identify common use cases and optimize them, do what you can to preserve time complexity guarantees of the operations performed on the source.

Reply:
I changed the implementation for updating sourceToTranslatedIndices by using splice, but I'm not sure if that addresses your comments. In terms of common use cases, should I be looking at how any of these arrays are sorted, or if the insertion is made at the start/end of the source array?

Reply:
Your implementation has a performance profile that always assumes the worst case. The goal of identifying common use cases is to optimize those use cases so that the average performance is improved. Put another way, this algorithm has a ```o(N)``` best run time and ```O(N)``` worst time. Identify common use cases to lower the bound on the best time from ```o(N)``` to something better, such as ```o(log(N))``` or ```o(1)```.